### ISSUE TITLE
Issue with results of MMDDrift and KSDrift

### ISSUE THREAD

**Issue by karthik-v-b** [SeldonIO/alibi-detect#545]
Hello,

I have used MMDDrift and KSDrift with tensorflow as backend to detect the drift detection with IMDB Movie review dataset, but when i run the script each time the Drift, p_val and distance  keeps on changing  for the same dataset.

Reference dataset consists of 500 positive reviews and 500 negative reviews
prediction dataset consists of 1000 reviews in which 900 are negative reviews

And to add on this i have passed the reference and prediction dataset as an pandas dataframe and i haven't used random method to get the data.

**Comment by ascillitoe** [comment]
Hi @karthik-v-b, there are a number of things that need to be considered if you want to get a detector behaving deterministically (i.e. giving repeatable `is_drift`, `p_val` and `distance` results). Before getting into this though, I should point out that usually you don't actually need a detector to be deterministic. Generally, it is the statistical behaviour of the detector over a large number of detection's that is important (i.e. the true and false positive rate, and the p-value distributions). Even if a detector is perfectly calibrated, we expect it to make a number of false positives (the FPR should actually approx. your chosen significance level `p_val`). 

As to how to get deterministic behaviour if you really want it, this will depend on the detector. 

### KS detector
For `KSDrift`, I would expect this to be deterministic with no extra effort on your part. If I run the imdb notebook (with minor edits for testing) I get:

```python
# initialize detector
cd = KSDrift(X_ref, p_val=.05, preprocess_fn=preprocess_fn, input_shape=(max_len,))

# Test predictions
labels = ['No!', 'Yes!']

preds_h0 = cd.predict(X_h0)
print('Drift? {}'.format(labels[preds_h0['data']['is_drift']]))
print('p-value: {}'.format(preds_h0['data']['p_val']))
print('distance: {}'.format(preds_h0['data']['distance']))

preds_h0 = cd.predict(X_h0)
print('Drift? {}'.format(labels[preds_h0['data']['is_drift']]))
print('p-value: {}'.format(preds_h0['data']['p_val']))
print('distance: {}'.format(preds_h0['data']['distance']))
```

returns:

```
Drift? No!
p-value: [0.722555   0.02565915 0.1338343  0.7590978  0.34099194 0.79439443
 0.85929435 0.1338343  0.31356168 0.34099194 0.7590978  0.19951835
 0.40047103 0.7590978  0.82795686 0.5726548  0.8879386  0.14833806
 0.50035924 0.6852314  0.18111965 0.7590978  0.99365413 0.9134755
 0.2406036  0.5726548  0.34099194 0.93558097 0.9134755  0.3699725
 0.40047103 0.96887016]
distance: [0.031 0.066 0.052 0.03  0.042 0.029 0.027 0.052 0.043 0.042 0.03  0.048
 0.04  0.03  0.028 0.035 0.026 0.051 0.037 0.032 0.049 0.03  0.019 0.025
 0.046 0.035 0.042 0.024 0.025 0.041 0.04  0.022]
Drift? No!
p-value: [0.722555   0.02565915 0.1338343  0.7590978  0.34099194 0.79439443
 0.85929435 0.1338343  0.31356168 0.34099194 0.7590978  0.19951835
 0.40047103 0.7590978  0.82795686 0.5726548  0.8879386  0.14833806
 0.50035924 0.6852314  0.18111965 0.7590978  0.99365413 0.9134755
 0.2406036  0.5726548  0.34099194 0.93558097 0.9134755  0.3699725
 0.40047103 0.96887016]
distance: [0.031 0.066 0.052 0.03  0.042 0.029 0.027 0.052 0.043 0.042 0.03  0.048
 0.04  0.03  0.028 0.035 0.026 0.051 0.037 0.032 0.049 0.03  0.019 0.025
 0.046 0.035 0.042 0.024 0.025 0.041 0.04  0.022]
```

The results are the same here, as expected. (note that you might sometimes get `Drift? Yes!`, since the false positive rate is likely not zero). 

### MMDDrift

This detector is not deterministic by default, since randomness is introduced by the permutation test used to obtain p-values. The `distance`'s returned should be repeatable, but the values of `p_val` and `is_drift` values are not. If you want deterministic behaviour you need to set the NumPy random seed before each prediction:

```python
cd = MMDDrift(X_ref, p_val=.05, preprocess_fn=preprocess_fn, 
              n_permutations=100, input_shape=(max_len,))

np.random.seed(0)
preds_h0 = cd.predict(X_h0)
print('Drift? {}'.format(labels[preds_h0['data']['is_drift']]))
print('p-value: {}'.format(preds_h0['data']['p_val']))
print('distance: {}'.format(preds_h0['data']['distance']))

preds_h0 = cd.predict(X_h0)
print('Drift? {}'.format(labels[preds_h0['data']['is_drift']]))
print('p-value: {}'.format(preds_h0['data']['p_val']))
print('distance: {}'.format(preds_h0['data']['distance']))

np.random.seed(0)
preds_h0 = cd.predict(X_h0)
print('Drift? {}'.format(labels[preds_h0['data']['is_drift']]))
print('p-value: {}'.format(preds_h0['data']['p_val']))
print('distance: {}'.format(preds_h0['data']['distance']))
```

should return:

```
Drift? No!
p-value: 0.45
distance: -8.344650268554688e-06
Drift? No!
p-value: 0.44
distance: -8.344650268554688e-06
Drift? No!
p-value: 0.45
distance: -8.344650268554688e-06
```

You should see the `distance` is the same in all three cases, but the second `p-value` is different since we didn't set the seed there.

### Other detectors

Most detectors behave in a similar way i.e. they should either always be deterministic, or will be as long as you set the NumPy (and sometimes `pytorch` and `tensorflow`) random seeds. 

A small number of detectors such as `ClassifierDrift` and `SpotTheDiffDrift` might still exhibit random behaviour even if the random seeds are set (https://github.com/SeldonIO/alibi-detect/issues/390). These detectors should still give reliable predictions despite this. 

If you are seeing different behaviour to that shown above, please do let us know and we will investigate further.
